### PR TITLE
[DPE-3266] - feat: in-place upgrades

### PR DESCRIPTION
**Pull request by marcoppenheimer**
#### `feat: add in-place upgrade support`
- Disabled the `resume-upgrade` functionality from the lib, to keep behavior the same between VM + K8s
- Although it works, upgrades on K8s needs to be manually tested more thoroughly before merging. NOTE - int test is skipped for now, as we don't have a published `stable` charm to test against

### CODE REVIEW

File: lib/charms/rolling_ops/v0/rollingops.py [hunk 1]
```diff
@@ -374,7 +374,7 @@ def _on_acquire_lock(self: CharmBase, event: ActionEvent):
             Lock(self).acquire()  # Updates relation data
             # emit relation changed event in the edge case where aquire does not
             relation = self.model.get_relation(self.name)
-            self.charm.on[self.name].relation_changed.emit(relation)
+            self.charm.on[self.name].relation_changed.emit(relation=relation, app=self.charm.app, unit=self.charm.unit)
```
Comment (zmraul) [inline lib/charms/rolling_ops/v0/rollingops.py:377]: Change on a lib without version change?

File: src/charm.py [hunk 2]
```diff
@@ -170,7 +180,10 @@ def _on_zookeeper_pebble_ready(self, event: EventBase) -> None:
         Handles case where workload has shut down due to failing `ruok` 4lw command and
         needs to be restarted.
         """
-        # FIXME: Will need updating when adding in-place upgrade support
+        # don't want to run default pebble ready during upgrades
+        if not self.upgrade_events.idle:
+            return
```
Comment (zmraul) [inline src/charm.py:185]: Should this be also deferred?

File: src/events/upgrade.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    def _on_zookeeper_pebble_ready_upgrade(self, _: EventBase) -> None:
+        """Handler for the `upgrade-charm` events handled during in-place upgrades."""
+        # ensure pebble-ready only fires after normal peer-relation-driven server init
+        if not self.charm.workload.alive or not self.charm.state.unit_server.started or self.idle:
+            return
+
+        try:
+            if self.charm.workload.healthy:
+                self.set_unit_completed()
```
Comment (zmraul) [inline src/events/upgrade.py:61]: Why is the `unit_completed` being called before `init_server`?

**Review by zmraul — APPROVED** [review]
LGTM. Just left some questions.

File: lib/charms/rolling_ops/v0/rollingops.py (hunk 1 above)
Comment (marcoppenheimer) [reply lib/charms/rolling_ops/v0/rollingops.py:377]: Ah, I forgot to publish it. Will fix in another PR, but I did merge it here - https://github.com/canonical/charm-rolling-ops/pull/5

File: src/charm.py (hunk 2 above)
Comment (marcoppenheimer) [reply src/charm.py:185]: If we're upgrading, we want to skip **this** `pebble_ready` call, and use the one in `events.upgrade`

File: src/events/upgrade.py (hunk 3 above)
Comment (marcoppenheimer) [reply src/events/upgrade.py:61]: I think it's because it **might** have self-healed. I don't think it actually will though, but just in case it comes up later on during our dev, that's why it's there.

Comment (deusebio) [reply src/events/upgrade.py:61]: Uhm, I also find it a bit weird and confusing. Can we just not do this? Would it be fair - even if the unit comes back online and self-heal, just do a init_server? I mean, even if it came back online, maybe the upgrade would trigger some updates of the server configuration. It seems a bit dangerous to just skip the init_server JUST because it is online

Also, What is the order of the events here? upgrade first and then pebble ready, or the other way around? Because if the

BTW, this seems a very general use-case, which is not Kafka specific. What people on PostgreSQL and MySQL do in these cases?

File: src/literals.py [hunk 4]
```diff
@@ -19,6 +19,15 @@
 JMX_PORT = 9998
 METRICS_PROVIDER_PORT = 7000
 
+DEPENDENCIES = {
+    "service": {
+        "dependencies": {},
+        "name": "zookeeper",
+        "upgrade_supported": "^3.5",
```
Comment (deusebio) [inline src/literals.py:26]: maybe 3.6? I don't think we have ever had any deployment before 3.6

**Review by deusebio — COMMENTED** [review]
Looks quite fine. I just have some suggestion to better fit in the new pattern we are using (since I feel the events/upgrade still has a bit too much business logic) but also I'm - similarly to @zmraul - a bit concerned with [this](https://github.com/canonical/zookeeper-k8s-operator/pull/60/files#r1463722602) for which I'd like to understand the use-case (and possible edge-cases) better

File: src/events/upgrade.py (hunk 3 above)
Comment (marcoppenheimer) [reply src/events/upgrade.py:61]: Maybe it's a good point. I'll remove it.

On the event order, for K8s it goes:
1. upgrade-charm
2. pebble-ready
3. start

File: src/charm.py [hunk 5]
```diff
@@ -156,8 +164,10 @@ def _on_cluster_relation_changed(self, event: EventBase) -> None:
         # only restart where necessary to avoid slowdowns
         # config_changed call here implicitly updates jaas + zoo.cfg
         if (
-            self.config_manager.config_changed() or self.state.cluster.switching_encryption
-        ) and self.state.unit_server.started:
+            (self.config_manager.config_changed() or self.state.cluster.switching_encryption)
+            and self.state.unit_server.started
+            and self.upgrade_events.idle
+        ):
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/charm.py:170]: It's now in the `events.upgrade.idle`, and we moved it down there.
For K8s, we do `idle` check differently, using `bool(self.upgrade_stack)` instead, so that logic is still there.

I moved it because we don't want to stop the updating of the quorum, just the restarts.

File: src/events/upgrade.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+            logger.info("Check failed: Quorum leader not found")
+            raise ClusterNotReadyError(message=default_message, cause="Quorum leader not found")
+        except ConnectionClosedError:
+            logger.info("Check failed: Unable to connect to the cluster")
+            raise ClusterNotReadyError(
+                message=default_message, cause="Unable to connect to the cluster"
+            )
+        except Exception as e:
+            logger.info(f"Check failed: Unknown error: {e}")
+            raise ClusterNotReadyError(message=default_message, cause="Unknown error")
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/events/upgrade.py:146]: Agreed, good idea. Will do!

File: src/events/upgrade.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    def idle(self) -> bool:
+        """Checks if cluster state is idle.
+
+        Returns:
+            True if cluster state is idle. Otherwise False
+        """
+        return not bool(self.upgrade_stack)
+
+    @cached_property
+    def client(self) -> ZooKeeperManager:
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/events/upgrade.py:88]: I agree in principle, but I'm not sure that constructing an object is repetition. We already have a library for the client itself, so different modules instantiate that library.

That said, if we move all the client checks to a manager (as per this comment - https://github.com/canonical/zookeeper-k8s-operator/pull/60/files/4f761ec1d123c691d2fb634ed4b9046501d29579#r1464336395), then we can move this out.

File: src/charm.py [hunk 8]
```diff
@@ -57,6 +59,12 @@ def __init__(self, *args):
         self.password_action_events = PasswordActionEvents(self)
         self.tls_events = TLSEvents(self)
         self.provider_events = ProviderEvents(self)
+        self.upgrade_events = ZKUpgradeEvents(
+            self,
+            dependency_model=ZooKeeperDependencyModel(
+                **DEPENDENCIES  # pyright: ignore[reportGeneralTypeIssues]
+            ),
+        )
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/charm.py:67]: During pod failover/rescheduling, we also get `upgrade-charm`.
During upgrades on K8s, the FIRST event we get is `upgrade-charm`.

We want the pod to failover normally, so we handle `charm._on_zookeeper_pebble_ready`.
However, we want to block this during an 'actual' upgrade upgrade. As such, we check for the existence of an `upgrade_stack` (which will have been set if we ran `pre-upgrade-check`). If so, the charm skips `charm._on_zookeeper_pebble_ready`, but DOES run `charm.upgrade_events._on_zookeeper_pebble_ready_upgrade` instead. 

We can observe to the same event multiple times, we skip the normal one and run the upgrade one in that case.

[the remark this answers is not in the document]
Comment (deusebio) [reply src/charm.py:67]: ah I see. I find it a bit confusing that on failover/rescheduling we run this event, but all right, that's the framework. Now I understand why we do this! 

I'm just wondering whether it is a good pattern to observe an event multiple times? I'm wondering whether we could centralize this logic in ONE upgrade method. Just saying this, because I would believe it is easier to maintain a code if we know that an event is processes in one place only, rather than having the possibility of the code related to one event to be scattered around

[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/charm.py:67]: So I agree on your last point. I think it's a bit funky and (apparently) confusing.

The lib should be looked at for that, but definitely with more attention :)

File: src/managers/quorum.py [hunk 9]
```diff
@@ -30,6 +32,58 @@ class QuorumManager:
...
+
+        return ZooKeeperManager(
+            hosts=active_hosts,
+            client_port=CLIENT_PORT,
+            username=admin_username,
+            password=admin_password,
+        )
+
+    @dataclass
+    class SyncStatus:
```
Comment (deusebio) [inline src/managers/quorum.py:50]: **minor** I find it weird to have classes within classes. I know we can, but I have not found it very often. I would just move it outside of the class. 

It is not a strong point just a bit of preference. But I'm more curious to understand what are the use cases when you would do that and why?

File: src/events/upgrade.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    service: DependencyModel
+
+
+class ZKUpgradeEvents(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "ZooKeeperCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+        self.substrate = self.charm.state.substrate
```
Comment (deusebio) [inline src/events/upgrade.py:44]: Why is this needed? There seems to be no usage of this in the class

**Review by deusebio — APPROVED** [review]
Nice! I have a couple of non-blocking comments! I'm quite curious about one thing, but the code would be fine either way. We can also continue the discussion offline and merge this PR

File: src/events/upgrade.py (hunk 10 above)
Comment (marcoppenheimer) [reply src/events/upgrade.py:44]: Good point! I'll remove it before merging.

File: src/managers/quorum.py (hunk 9 above)
Comment (marcoppenheimer) [reply src/managers/quorum.py:50]: I agree, it usually seems a bit clunky to nest classes/functions.

My motivation here was to avoid endless tiny types, which makes mentally 'parsing' the code much more taxing. I felt that here, typing the Status to `SyncStatus` specifically, something that's referenced through `QuorumManager.SyncStatus` seemed logical to me.

I'd actually be fine moving it out if we need to, but that was the reasoning at least.

File: src/events/upgrade.py (hunk 10 above)
Comment (marcoppenheimer) [reply src/events/upgrade.py]: I was wrong to do this. We need the substrate, else the `pre-upgrade-check` defaults to "vm", and attempts to run a not-implemented `build_upgrade_stack`.

Will fix.
